models/src/main/scala/coop/rchain/models/SortedParMap.scala

```diff
@@ -1,29 +1,54 @@
 package coop.rchain.models
 
-import scala.collection.MapLike
 import coop.rchain.models.rholang.sorter.ordering._
 
-class SortedParMap private (private val ps: Map[Par, Par])
-    extends Map[Par, Par]
-    with MapLike[Par, Par, SortedParMap] {
+import scala.collection.GenTraversableOnce
+import scala.collection.immutable.ListMap
 
-  lazy val sortedMap: Map[Par, Par] = ps.sort.toMap
+final class SortedParMap private (ps: Map[Par, Par]) extends Iterable[(Par, Par)] {
 
-  override def empty: SortedParMap = SortedParMap(Map.empty[Par, Par])
+  // TODO: Merge `sortedMap` and `sortedMapLookup` into one VectorMap once available
+  lazy val sortedMap: ListMap[Par, Par]   = ListMap(ps.sort: _*)
+  lazy val sortedMapLookup: Map[Par, Par] = ps.sort.toMap
 
-  override def get(key: Par): Option[Par] = sortedMap.get(key)
+  def +(kv: (Par, Par)): SortedParMap = SortedParMap(sortedMap + kv)
 
-  override def iterator: Iterator[(Par, Par)] = sortedMap.toIterator
+  def ++(kvs: GenTraversableOnce[(Par, Par)]): SortedParMap = SortedParMap(sortedMap ++ kvs)
 
-  override def +[V1 >: Par](kv: (Par, V1)): Map[Par, V1] = sortedMap + kv
+  def -(key: Par): SortedParMap = SortedParMap(sortedMap - key)
```

Comment:
the key itself may be not sorted. Take e.g. 

//ks == Sortable[Par].sort(ku), ks != ku
val m = SortedMap(ku -> 1)
val m2 = m - ku
//m2 == m, while it shouldn't
//this is because:
//m ~= Map(ks -> 1)

Reply:
It should be enough to only sort the keys and not values when adding and removing elements from the Map and sort everything else when checking equality and before serialization.

Reply:
Nopes, you'd also have to values when accessing elements. Also, since the types don't allow us to discern sorted from non-sroted ones, I want to keep the invariant of 'everything inside is sorted'.

Reply:
Then sort when `EMap` are reduced/interpreted/evaled.

Reply:
I am not sure if your approach is correct. You make internals leak to the user.
```scala
sortedMap.add(key, value).get(key) != value
```
because of sortedness. It is your call though.

Reply:
Thanks for that remark! Indeed, that's undesirable.

Again though, the types are failing us here, because we've failed to use them in the first place. It is not an implementation detail of the SortedMap. Quite the contrary, it's the `raison d'etre` for it ;) SortedMap is used to represent a sorted map term. Such term has a unique canonical shape and this shape should be used here :)

Previously, the SortedMap had broken equals because `get` returned unsorted elements. But `SortedMap(1-> [2, 3])` should equal `SortedMap(1 -> [3, 2])`. Otherwise the latter map does not represent a sorted term :)

Let me know if that convinces you to the changed behavior :)

re: your proposal about postponing the sorting - I'd rather not reason about subsequent potential usages of the SortedMap and instead make all of them behave consistently. I'm not good at such reasoning. The types are great at that - but we haven't invited them :hits_blunt:

Reply:
> But `SortedMap(1-> [2, 3])` should equal `SortedMap(1 -> [3, 2])`

Well, this example is not correct because `[2,3]` is not equal to `[3,2]` but `SortedMap(1 -> p | q)` is equal to `SortedMap(1 -> q | p)`. 
It also depends which algorithm is used for equality check. For example, if you use Rholang's `==` then `p == q != q == p`. 

Anyway, I think that users should not observe the sortedness property of a Map. The sorting was added for the serialization sake and has now bubbled up up to the user.

IMO there should be an internal, sorted representation for purposes of `contains` and `equals` but `get` should return unsorted value.

Reply:
> if you use Rholang's == then `p == q != q == p`

WDYM? Rholang's `==` is not symmetric?

Also, I'd totally expect `{ p | q } == { q | p }` in rholang.

> sorting was added for the serialization sake 

AFAIK it's also a crucial concept behind all equality and pattern matching (pattern matching is just equality whenever the pattern has no free variables or wildcards).

Does that change your view of things?

Reply:
I agree with @goral09 that `SortedMap(1-> [2, 3])` should not be equal to `SortedMap(1-> [3, 2])` since a List should keep its order.
AFAIK this is also how it is implemented right now.

Reply:
I also agree with him on that :) I really meant the example he's given.

`SortedMap(1 -> q | p) == SortedMap(1 -> p | q)` should be true in current implementation, and in general.

Reply:
> IMO there should be an internal, sorted representation for purposes of contains and equals but get should return unsorted value.

If this was true, the following would be false:

```
Map(1 -> 3 | 2).get(1) == 3 | 2
```

That is, if I'm able to create an unsorted (sub)term in the first place.
The thing is, at times the terms are being made by Scala code other than the normalizer (reductions, substitutions, `Method` calls) and there's no easy way to guarantee their outputs are sorted.

Since all the terms that can be written are 'born' sorted, there's no error in only returning sorted terms from the map.

Since the dynamically created terms (e.g. method call results) can't be guaranteed to be sorted, sorting them on entry might be costly, but is safer.

Reply:
That's what I said. Rholang's `==` (equality sign) doesn't support these properties (like spatial matcher does). It just uses Java equality and I think this is something users should be aware of.

Reply:
Since there's no way to have unsorted terms written in soruce code, and all rholang's internals should be returning sorted terms, there's nothing for the users to be aware of. 

*Throws chair.*

Reply:
> Since there's no way to have unsorted terms written in soruce code,

That's not true. IIRC sorting happens only after substitution. It means that if you write down concrete, unsorted term then it stays unsorted.

Reply:
Nopes, buildNormalizedTerm also sorts (see buildPar) :) I think we can finally agree :)

Reply:
Doesn't it sort only the outermost elements? Not the inner ones? Like elements of the collection for example.

Reply:
Nopes, sorting is recursive.

Reply:
Then why do we have to sort at all when interacting with Map if everything is already sorted?

Reply:
As I said, we don't have such guarantees when getting dynamically-generated
terms (subtitutions, Method results). I'd like the map to sort again until
we find a way to encode the sortedness in types or otherwise guarantee
sortedness of dynamically generated terms. Unless we prove that this is a
huge performance drag (i.e. have an e2e benchmark for that).